docs/design/sql/05-ddl.md

```diff
@@ -0,0 +1,306 @@
+3. [DROP EXTERNAL TABLE](#3-drop-external-table) -- remove external table
+
+## 1 CREATE EXTERNAL TABLE
+
+`CREATE EXTERNAL TABLE` -- define a new external table
+
+### 1.1 Synopsis
+
+```sql
+CREATE [OR REPLACE] EXTERNAL TABLE [IF NOT EXISTS] table_name [ (
+ column_name column_type [ EXTERNAL NAME column_external_name ] [, ... ]
+) ]
+TYPE table_type
+[ OPTIONS ( table_parameter table_parameter_value [, ... ] ) ]
+```
+
+### 1.2 Description
+
+`CREATE EXTERNAL TABLE` registers a virtual table that references an external storage system. For some storage systems,
+it does not create a physical entity until a write occurs. Column list is optional, as for some storage systems and
+serialization formats, it is possible to infer it.
+
+The table is created in the `public` schema. The name of the table must be distinct from the name of any other table in
+the schema.
+
+### 1.3 Parameters
+
+**OR REPLACE**
+
+Replace table definition if one with the same name already exists.
+
+**IF NOT EXISTS**
+
+Do not fail if a table with the same name already exists.
+
+**table_name**
+
+The unqualified name of the table to be created.
+
+**column_name**
+
+The name of the column in the table.
+
+**column_type**
+
+The data type of the column. For more information on the supported data types, refer to [type system
+document](01-type-system.md).
+
+**column_external_name**
+
+The path to the field in the external storage system. Used to link a column to physical field under a different name
+(i.e. to be able to access both key and value `IMap` fields having same name).
+
+**table_type**
+
+The external storage system identifier. Supported table types are listed in Table 1. Not all storage systems implement
+full set of operations, so for instance you might be able to `SELECT` from a given table but not `INSERT` into it.
+
+*Table 1: CREATE EXTERNAL TABLE supported table types*
+
+| Table type | Description |
+|---|---|
+| com.hazelcast.IMap | Table backed by Hazelcast `IMap` |
+| com.hazelcast.ReplicatedMap | Table backed by Hazelcast `ReplicatedMap` |
+| com.hazelcast.File | Table backed by files in a directory |
+
+**table_parameter**
+
+Parameter specific to the external storage system - connection string, serialization format etc.
+
+**table_parameter_value**
+
+Parameter value as a string literal.
+
+### 1.4 Examples
+
+- Create an external table referencing an `IMap` with inferred schema.
+
+  ```sql
+  CREATE EXTERNAL TABLE persons
+  TYPE "com.hazelcast.IMap"
+  OPTIONS (
+    "serialization.key.format" 'java',
+    "serialization.key.java.class" 'java.lang.Integer',
+    "serialization.value.format" 'java',
+    "serialization.value.java.class" 'java.lang.String'
+  )
+  ```
+
+  Column names and types are inferred as:
+
+  ```sql
+  (
+    __key INT,
+    this VARCHAR
+  )
+  ```
+
+- Create an external table referencing an `IMap` with explicit schema.
+
+  ```sql
+  CREATE EXTERNAL TABLE persons (
+    id INT EXTERNAL NAME __key,
```

Comment:
Does any other connector except for IMap/ReplicatedMap requires `EXTERNAL NAME`?

Reply:
Kafka might use it in similar way (to disambiguate key and value fields). Moreover it might be simply used to operate with different names on SQL level (not sure how useful that might be though).

Reply:
For maps and kafka it's required only for key fields. If not present, we assume a value field. But this is type-specific, for example, external JDBC table will never require it, but could be used to rename a column.

Reply:
Got it. Another question - why do we call it `EXTERNAL NAME`? In terms of Hazelcast, this is a "path", not something external. Basically, we create *aliases* for paths.

Reply:
External in a sense that it refers to something outside of SQL - i.e. a field in java object in an IMap. Sometimes that would be simply an alias (if you want to just rename) but sometimes a path (if you want to disambiguate). I'm not happy with `external name` but could not find a better name...

Reply:
It is not a blocker definitely for the whole activity, just sounds a bit strange. Do any other NoSQL products with SQL support have something like this? Maybe we can borrow ideas from them.

A couple of ideas from my side:
1. Maybe we can define a custom function to extract the specific field from the key/value. E.g. `SELECT FIELD('name', __key), name FROM Person` - this first field is taken from the key, the second one is taken from the value
1. Provided that the conflicts are expected to be rare, we may use a custom property instead with a column name attached. E.g. `CREATE TABLE ... OPTIONS path.name=__key.name`. The advantage of this is that the basic syntax is not concerned with connector-specific stuff